awx/ui_next/src/screens/Template/shared/JobTemplateForm.jsx

```diff
@@ -36,22 +50,113 @@ class JobTemplateForm extends Component {
       playbook: '',
       summary_fields: {
         inventory: null,
+        labels: { results: [] },
       },
     },
   };
 
   constructor(props) {
     super(props);
-
     this.state = {
+      hasContentLoading: true,
+      contentError: false,
+      loadedLabels: [],
+      newLabels: [],
+      removedLabels: [],
       inventory: props.template.summary_fields.inventory,
     };
+    this.handleNewLabel = this.handleNewLabel.bind(this);
+    this.loadLabels = this.loadLabels.bind(this);
+    this.disassociateLabel = this.disassociateLabel.bind(this);
+  }
+
+  componentDidMount() {
+    this.loadLabels(QSConfig);
+  }
+
+  async loadLabels(QueryConfig) {
+    this.setState({ contentError: null, hasContentLoading: true });
+    let loadedLabels;
+    try {
+      const { data } = await LabelsAPI.read(QueryConfig);
+      loadedLabels = [...data.results];
+      if (data.next && data.next.includes('page=2')) {
+        const {
+          data: { results },
+        } = await LabelsAPI.read({
+          page: 2,
+          page_size: 200,
+          order_by: 'name',
+        });
+        loadedLabels = loadedLabels.concat(results);
+      }
+      this.setState({ loadedLabels });
+    } catch (err) {
+      this.setState({ contentError: err });
+    } finally {
+      this.setState({ hasContentLoading: false });
+    }
+  }
+
+  handleNewLabel(label) {
+    const { newLabels } = this.state;
+    const { template } = this.props;
+    const isIncluded = newLabels.some(newLabel => newLabel.name === label.name);
+    if (isIncluded) {
+      const filteredLabels = newLabels.filter(
+        newLabel => newLabel.name !== label
+      );
+      this.setState({ newLabels: filteredLabels });
+    } else if (typeof label === 'string') {
+      this.setState({
+        newLabels: [
+          ...newLabels,
+          {
+            name: label,
+            organization: template.summary_fields.inventory.organization_id,
+          },
+        ],
+      });
+    } else {
+      this.setState({
+        newLabels: [
+          ...newLabels,
+          { name: label.name, associate: true, id: label.id },
+        ],
+      });
+    }
+  }
+
+  disassociateLabel(label) {
```

Comment:
I think we usually keep the `disassociate` / `associate` words for things that make network calls. Might this be better named as `removeLabel` or something like that?

Reply:
👍